File: prefPane/id-updater-helper.m
/*
 * id-updater
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */

#import "update.h"

#define UPDATER_ID @"ee.ria.id-updater"

@interface Updater : Update <UpdateDelegate> {
    NSString *path;
}
@end

@implementation Updater
- (id)initWithPath:(NSString *)_path {
    if (self = [super initWithDelegate:self]) {
        path = _path;
        NSLog(@"Installed %@: %@", path, self.baseversion);
        [self request];
    }
    return self;
}

#pragma mark - Update Delegate

- (void)didFinish:(NSError *)error {
    if (error) {
        NSLog(@"Error: %@", error.localizedDescription);
    }
    exit(0);
}

- (void)message:(NSString *)message {
    NSLog(@"%@", message);
    [[NSTask launchedTaskWithLaunchPath:@"/usr/bin/open" arguments:@[path]] waitUntilExit];
}

- (void)updateAvailable:(NSString *)available filename:(NSString *)filename {
    NSLog(@"Update available %@ %@", available, filename);
    [[NSTask launchedTaskWithLaunchPath:@"/usr/bin/open" arguments:@[path]] waitUntilExit];
}
@end

#pragma mark - Main

int main(int argc, const char * argv[])
{
    if (argc != 2)
        return 1;

    @autoreleasepool {
        if (strcmp(argv[1], "-task") == 0) {
            Updater *updater __attribute__((unused)) = [[Updater alloc] initWithPath:[NSString stringWithFormat:@"%s/../../..", argv[0]].stringByStandardizingPath];
            [NSRunLoop.mainRunLoop run];
            return 0;
        }

        NSString *PATH = (@"~/Library/LaunchAgents/ee.ria.id-updater.plist").stringByStandardizingPath;
        [NSFileManager.defaultManager createDirectoryAtPath:(@"~/Library/LaunchAgents").stringByStandardizingPath withIntermediateDirectories:YES attributes:nil error:nil];
        if (strcmp(argv[1], "-remove") == 0) {
            [[NSTask launchedTaskWithLaunchPath:@"/bin/launchctl" arguments:@[@"unload", @"-w", PATH]] waitUntilExit];
            NSError *error;
            [[NSFileManager defaultManager] removeItemAtPath:PATH error:&error];
            return 0;
        }

        NSDateComponents *components = [[NSCalendar currentCalendar]
                                        components:NSCalendarUnitHour|NSCalendarUnitMinute|NSCalendarUnitWeekday|NSCalendarUnitDay
                                        fromDate:[NSDate date]];
        NSNumber *hour = @(components.hour);
        NSNumber *minute = @(components.minute);
        NSNumber *weekday = @(components.weekday);
        NSNumber *day = @(components.day);
        NSDictionary *settings = nil;
        if (strcmp(argv[1], "-daily") == 0) {
            settings = @{@"Label": UPDATER_ID,
                         @"ProgramArguments": @[@(argv[0]), @"-task"],
                         @"StartCalendarInterval": @{@"Hour": hour, @"Minute": minute}};
        } else if (strcmp(argv[1], "-weekly") == 0) {
            settings = @{@"Label": UPDATER_ID,
                         @"ProgramArguments": @[@(argv[0]), @"-task"],
                         @"StartCalendarInterval": @{@"Hour": hour, @"Minute": minute, @"Weekday": weekday}};
        } else if (strcmp(argv[1], "-monthly") == 0) {
            settings = @{@"Label": UPDATER_ID,
                         @"ProgramArguments": @[@(argv[0]), @"-task"],
                         @"StartCalendarInterval": @{@"Hour": hour, @"Minute": minute, @"Day": day}};
        } else {
            return 0;
        }
        NSError *error;
        NSData *data = [NSPropertyListSerialization dataWithPropertyList:settings format:NSPropertyListXMLFormat_v1_0 options:0 error:&error];
        [data writeToFile:PATH atomically:YES];
        [[NSTask launchedTaskWithLaunchPath:@"/bin/launchctl" arguments:@[@"load", @"-w", PATH]] waitUntilExit];
    }
    return 0;
}

File: prefPane/id-updater-helper.swift
/*
 * id-updater
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 */
 
import Foundation

@main
class Updater: Update, UpdateDelegate, @unchecked Sendable {
    private let path: String

    init(path: String) {
        self.path = path
        super.init()
        delegate = self
        print("Installed \(path): \(self.baseversion ?? "")")
        request()
    }

    static func launch(_ path: String, arguments: [String]) -> Int32 {
        let task = Process()
        task.launchPath = path
        task.arguments = arguments
        task.launch()
        task.waitUntilExit()
        return task.terminationStatus
    }

    // MARK: - Update Delegate

    func didFinish(_ error: Error?) {
        if let error = error {
            print("Error: \(error.localizedDescription)")
        }
        exit(0)
    }

    func message(_ message: String) {
        print(message)
        _ = Updater.launch("/usr/bin/open", arguments: [path])
    }

    func updateAvailable(_ available: String, filename: String) {
        print("Update available \(available) \(filename)")
        _ = Updater.launch("/usr/bin/open", arguments: [path])
    }

    static func main() {
        let arguments = CommandLine.arguments
        guard arguments.count == 2 else { 
            print("Usage: \(arguments[0]) [-task | -remove | -daily | -weekly | -monthly]")
            exit(1)
        }

        let directoryPath = ("~/Library/LaunchAgents" as NSString).expandingTildeInPath
        let PATH = directoryPath + "/ee.ria.id-updater.plist"
        let components = Calendar.current.dateComponents([.hour, .minute, .weekday, .day], from: Date())
        let schedule: [String: Any]

        switch arguments[1] {
        case "-task":
            _ = Updater(path: NSString(string: "\(Bundle.main.executablePath ?? arguments[0])/../../..").standardizingPath)
            return RunLoop.main.run()
        case "-remove":
            let result = Updater.launch("/bin/launchctl", arguments: ["unload", "-w", PATH])
            try? FileManager.default.removeItem(atPath: PATH)
            exit(result)
        case "-daily":
            schedule = ["Hour": components.hour!, "Minute": components.minute!]
        case "-weekly":
            schedule = ["Hour": components.hour!, "Minute": components.minute!, "Weekday": components.weekday!]
        case "-monthly":
            schedule = ["Hour": components.hour!, "Minute": components.minute!, "Day": components.day!]
        default:
            print("Invalid argument: \(arguments[1])")
            exit(1)
        }

        let settings: [String: Any] = [
            "Label": "ee.ria.id-updater",
            "ProgramArguments": [arguments[0], "-task"],
            "StartCalendarInterval": schedule
        ]
        let plistData = try? PropertyListSerialization.data(fromPropertyList: settings, format: .xml, options: 0)
        try? FileManager.default.createDirectory(atPath: directoryPath, withIntermediateDirectories: true, attributes: nil)
        FileManager.default.createFile(atPath: PATH, contents: plistData, attributes: nil)
        exit(Updater.launch("/bin/launchctl", arguments: ["load", "-w", PATH]))
    }
}
